# 배열이란?

배열을 사용하는 기본 알고리즘 알아보기. 코테 연습 문제 중 Lv. 0에 해당하는 알고리즘이 주로 포함된 듯

## 배열 원소의 최댓값 구하기

In [2]:
def max_of(a):
    maximum = a[0]
    for i in range(1, len(a)):
        if a[i] > maximum:
            maximum = a[i]
    return maximum

위와 같이 배열 원소를 하나씩 차례로 주목하여 살펴보는 방식을 알고리즘 용어로 **스캔**(scan) 또는 트래버스(traverse)라고 함.

## 배열 원소의 최댓값을 구하는 함수 및 프로그램 구현하기

[Do it! 실습 2-2]: 시퀀스 원소의 최댓값 출력하기

함수는 다음과 같이 정의할 수 있으며, 이는 임포트를 위해 작업 디렉토리에 복사될 예정

In [4]:
from typing import Any, Sequence

def max_of(a: Sequence) -> Any:
    """시퀀스형 a 원소의 최댓값을 반환""" # 이를 함수 어노테이션이라 함
    maximum = a[0]
    for i in range(1, len(a)):
        if a[i] > maximum:
            maximum = a[i]
    return maximum

if __name__ == '__main__':
    print('배열의 최댓값을 구합니다.')
    
    # step 1: 배열의 원소 수 입력받기
    num = int(input('원소 수를 입력하세요.: '))
    x = [None] * num # 원소 수가 num인 리스트 생성
    
    # step 2: 배열의 값 입력받기
    for i in range(num):
        x[i] = int(input(f'x[{i}]값을 입력하세요.: '))
        
    # step 3: 최댓값 구하기 및 반환하기
    print(f'최댓값은 {max_of(x)}입니다.')

배열의 최댓값을 구합니다.
원소 수를 입력하세요.: 5
x[0]값을 입력하세요.: 172
x[1]값을 입력하세요.: 153
x[2]값을 입력하세요.: 192
x[3]값을 입력하세요.: 140
x[4]값을 입력하세요.: 165
최댓값은 192입니다.


Any의 경우 제약이 없는 임의의 자료형을 의미하며, Sequence의 경우 시퀀스형(sequence type)을 의미함. 시퀀스형에는 list, bytearray, string, tuple, bytes형이 있음

위에 따라 `max_of()`함수는 다음과 같이 정의할 수 있음
 * 건네받는 매개변수 a의 자료형은 Sequence임
 * 반환하는 것은 임의의 자료형임
 
 `max_of()`의 특성
 
 * 함수 안에서는 배열 a의 원솟값을 변경하지 않음
 * 호출하는 쪽이 넘겨주는 자료형은 시퀀스형이면 어떤 것이든 상관없음
 * 인수의 원소를 비교 연산자 >로 값을 비교할 수 있으면 다른 형과 섞여 있어도 됨
 * 최댓값의 원소가 int이면 int형을 반환, float이면 float형 값을 반환함

## 재사용할 수 있는 모듈 작성하기

파이썬에서는 하나의 스크립트 프로그램을 **모듈**(module)이라고 함. 확장자를 뺀 파일 이름 자체가 모듈 이름임. 위 프로그램을 max.py로 만들어 저장하자. 위에서 `__name__` 이 `'__main__'`과 같은지 판단하는데, 여기서 피연산자 `__name__`은 모듈 이름을 나타내는 변수이고 작성 규칙은 다음과 같음

* 스크립트 프로그램이 직접 실행될 때 `__name__`은 `'__main__'`임
* 스르립트 프로그램이 임포트될 때 변수 `__name__`은 원래의 모듈 이름임

파이썬에서는 모듈 또한 객체임. 모듈을 프로그램이 처음 임포트되는 시점에 모듈 객체가 생성되면서 초기화되는 구조임.

## 모듈 테스트하기

우선 위에서 작성한 파일을 외부에 모듈로 작성하자

### 입력받을 때 원소 수를 결정하기

[Do it! 실습 2-3] 배열 원소의 최댓값을 구해서 출력하기(원솟값을 입력받음)

In [5]:
from max import max_of

print('배열의 최댓값을 구합니다.')
print('주의: "End"를 입력하면 종료합니다.')

number = 0
x = []

while True:
    s = input(f'x[{number}]값을 입력하세요.: ')
    if s == 'End':
        break
    x.append(int(s))
    number += 1
    
print(f'{number}개를 입력했습니다.')
print(f'최댓값은 {max_of(x)}입니다.')

배열의 최댓값을 구합니다.
주의: "End"를 입력하면 종료합니다.
x[0]값을 입력하세요.: 15
x[1]값을 입력하세요.: 72
x[2]값을 입력하세요.: 64
x[3]값을 입력하세요.: 7
x[4]값을 입력하세요.: End
4개를 입력했습니다.
최댓값은 72입니다.


위에서 숫자가 아닌 원소가 입력되면 분명 에러가 날 것이며, float형의 데이터는 int 부분만 추려져서 데이터가 변형됨

### 배열의 원솟값을 난수로 결정하기

[Do it! 실습 2-4] 배열 원소의 최댓값을 구해서 출력하기(원솟값을 난수로 생성)

In [6]:
import random
from max import max_of

print('난수의 최댓값을 구합니다.')
num = int(input('난수의 개수를 입력하세요.: '))
lo = int(input('난수의 최솟값을 입력하세요.: '))
hi = int(input('난수의 최댓값을 입력하세요.: '))
x = [None] * num # 원소 수가 num인 리스트를 생성

for i in range(num):
    x[i] = random.randint(lo, hi) #lo 이상 hi 이하인 정수 난수를 반환
    
print(f'{(x)}')
print(f'이 가운데 최댓값은 {max_of(x)}입니다.')

난수의 최댓값을 구합니다.
난수의 개수를 입력하세요.: 5
난수의 최솟값을 입력하세요.: 10
난수의 최댓값을 입력하세요.: 99
[99, 90, 18, 92, 56]
이 가운데 최댓값은 99입니다.


### 튜플, 문자열, 문자열 리스트의 최댓값 구하기

[Do it! 실습 2-4] 배열 원소의 최댓값을 구해서 출력하기(튜플, 문자열, 문자열 리스트)

In [7]:
from max import max_of

t = (4, 7, 5.6, 2, 3.14, 1)
s = 'string'
a = ['DTS', 'AAC', 'FLAC']

print(f'{t}의 최댓값은 {max_of(t)}입니다.')
print(f'{s}의 최댓값은 {max_of(s)}입니다.')
print(f'{a}의 최댓값은 {max_of(a)}입니다.')

(4, 7, 5.6, 2, 3.14, 1)의 최댓값은 7입니다.
string의 최댓값은 t입니다.
['DTS', 'AAC', 'FLAC']의 최댓값은 FLAC입니다.


In [9]:
b = ['DTS', 'AAC', 'FLA']
c = ['DTS', 'AACDDDDD', 'FLA']

print(f'{b}의 최댓값은 {max_of(b)}입니다.')
print(f'{c}의 최댓값은 {max_of(c)}입니다.')

['DTS', 'AAC', 'FLA']의 최댓값은 FLA입니다.
['DTS', 'AACDDDDD', 'FLA']의 최댓값은 FLA입니다.


문자열 비교의 경우, 문자열의 첫 문자의 문자 코드가 크면 큰 것으로 간주함.

## 보충수업: 리스트와 튜플 2

### 동일성(indentity) 판단하기

In [11]:
lst1 = [1,2,3,4,5]
lst2 = [1,2,3,4,5]

lst1 is lst2

False

둘은 값은 같지만, 식별번호는 다름. 다른 객체를 참조하고 있기 때문. 단 그 객체의 값은 같음

### 리스트, 튜플의 대입

In [12]:
list1 = [1,2,3,4,5]
lst2 = lst1
lst1 is lst2

True

In [13]:
lst1[2] = 9
lst1, lst2

([1, 2, 9, 4, 5], [1, 2, 9, 4, 5])

위처럼 따로 생성한 것이 아닌, lst1을 lst2에 대입하면 같은 식별번호의 객체를 참조하게 되어, 한 쪽에서 값을 변경하면 다른 객체에서도 동일하게 변경 사항이 적용된 것으로 보임. 둘 다 적용된 것이 아니라 실제로는 lst1이라는 변수를 통해 객체의 값을 변경한 것이고 lst1, lst2가 동일 객체를 참조하여 같은 값이 출력되는 것을 알 수 있음.

### 리스트 스캔

튜플의 스캔도 아래와 동일함

프로그램 4종류

 * 실습 2C-1: 원소 수를 len() 함수로 미리 알아내서, 0에서 원소 수 - 1까지 반복
 * 실습 2C-2: 인덱스와 원소를 짝지어 enumerate() 함수로 반복해서 꺼냄
 * 실습 2C-3: 실습 2C-2와 같지만 카운트를 1부터 시작
 * 실습 2C-4: 인덱스를 사용하지 않고 in을 사용해 원소를 순서대로 꺼냄

In [14]:
x = ['John', 'George', 'Paul', 'Ringo']

[Do it! 실습 2C-1] 원소 수를 len() 함수로 미리 알아내서, 0에서 원소 수 - 1까지 반복

In [15]:
for i in range(len(x)):
    print(f'x[{i}] = {x[i]}')

x[0] = John
x[1] = George
x[2] = Paul
x[3] = Ringo


[Do it! 실습 2C-2] 인덱스와 원소를 짝지어 enumerate() 함수로 반복해서 꺼냄

In [16]:
for i, name in enumerate(x):
    print(f'x[{i}] = {name}')

x[0] = John
x[1] = George
x[2] = Paul
x[3] = Ringo


[Do it! 실습 2C-3] 실습 2C-2와 같지만 카운트를 1부터 시작

In [18]:
for i, name in enumerate(x, 1):
    print(f'{i}번째 = {name}')

1번째 = John
2번째 = George
3번째 = Paul
4번째 = Ringo


[Do it! 실습 2C-4] 인덱스를 사용하지 않고 in을 사용해 원소를 순서대로 꺼냄

In [20]:
for name in x:
    print(name)

John
George
Paul
Ringo


** 이터러블

문자열, 리스트, 튜플, 집합, 딕셔너리 등의 자료형 객체는 모두 이터러블(iterable, 반복 가능)하다는 공통점이 있음. 이는 원소를 하나씩 꺼내는 구조이며, 이터러블 객체를 내장 함수인 iter()의 인수로 전달하면, 그 객체에 대한 이터레이터(iterator)를 반환함.이는 데이턴의 나열을 표현하는 객체로 이를 통해 원소를 순차적으로 꺼낼 수 있음

In [23]:
for i in iter(x):
    print(i)

John
George
Paul
Ringo


## 배열 원소를 역순으로 정렬하기

[Do it! 실습 2-6] 뮤터블 시퀀스 원소를 역순으로 정렬

In [31]:
from typing import Any, MutableSequence

def reverse_array(a: MutableSequence) -> Any:
    for i in range(len(a)//2):
        a[i], a[len(a)-1-i] = a[len(a)-1-i], a[i]

if __name__ == '__main__':
    print('배열 원소를 역순으로 정렬합니다.')
    nx = int(input('원소 수를 입력하세요.: '))
    x = [None] * nx
    
    for i in range(nx):
        x[i] = int(input(f'x[{i}]값을 입력하세요.: '))
        
    reverse_array(x)
    
    print('배열 원소를 역순으로 정렬했습니다.')
    
    for i in range(nx):
        print(f'x[{i}] = {x[i]}')

배열 원소를 역순으로 정렬합니다.
원소 수를 입력하세요.: 7
x[0]값을 입력하세요.: 2
x[1]값을 입력하세요.: 5
x[2]값을 입력하세요.: 1
x[3]값을 입력하세요.: 3
x[4]값을 입력하세요.: 9
x[5]값을 입력하세요.: 6
x[6]값을 입력하세요.: 7
배열 원소를 역순으로 정렬했습니다.
x[0] = 7
x[1] = 6
x[2] = 9
x[3] = 3
x[4] = 1
x[5] = 5
x[6] = 2


다른 방법은?

In [37]:
x.reverse()
x

[7, 6, 9, 3, 1, 5, 2]

In [38]:
list(reversed(x))

[2, 5, 1, 3, 9, 6, 7]

위의 경우 reversed() 함수는 역순으로 정렬하는 이터러블 객체를 생성함. 따라서 이를 정렬하여 반환하려면 list() 함수에 전달하여 새로운 리스트를 생성해야 함

## 기수 변환하기(n진수 구하기)

기수는 수를 나타내며, 서수는 순서를 나타낸다는 차이가 있음. 10진수를 n진수로 나타내기 위해서는 10진수를 n으로 계속해서 나누고 더 이상 나눌 수 없을 때 중단함. 나눌 때마다 생성된 나머지의 수를 역순으로 하면 n진수로 나타낼 수 있음.

[Do it! 실습 2-7] 10진수 정숫값을 입력받아 2~36진수로 변환하여 출력하기

In [41]:
def card_conv(x: int, r: int) -> str:
    '''정숫값 x를 r진수로 변환한 뒤 그 수를 나타내는 문자열을 반환'''
    
    d = ''
    dchar = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ'
    
    while x > 0:
        d += dchar[x % r]
        x //= r
        
    return d[::-1]

if __name__ == '__main__':
    print('10진수를 n진수로 변환합니다.')
    
    while True:
        while True:
            no = int(input('변환할 값으로 음이 아닌 정수를 입력하세요.: '))
            if no > 0:
                break
                
        while True:
            cd = int(input('어떤 진수로 변환할까요?: '))
            if 2 <= cd <= 36:
                break
        
        print(f'{cd}진수로는 {card_conv(no, cd)}입니다.')
        
        retry = input("한 번 더 변환할까요?(Y ... 예/ N ... 아니오): ")
        
        if retry in {'N', 'no'}:
            break

10진수를 n진수로 변환합니다.
변환할 값으로 음이 아닌 정수를 입력하세요.: 29
어떤 진수로 변환할까요?: 2
2진수로는 11101입니다.
한 번 더 변환할까요?(Y ... 예/ N ... 아니오): N


## 보충 수업 2-6 함수 사이에 인수 주고 받기

In [44]:
def sum_1ton(n):
    '''1부터 n까지 정수의 합'''
    s = 0
    while n > 0:
        s += n
        n -= 1
    return s

x = int(input('x의 값을 입력하세요.: '))
print(f'1부터 {x}까지 정수의 합은 {sum_1ton(x)}입니다.')

x의 값을 입력하세요.: 5
1부터 5까지 정수의 합은 15입니다.


sum_1ton() 함수의 실행 과정에서 n의 값은 5 -> 4 -> 3 -> 2 -> 1 -> 0 으로 1씩 감소함. 그래서 함수가 종료될 때 n의 값은 0임. 그러나 x의 값은 그대로 유지되고 있음. 그렇다면 복사되는 것일까?

X 아님. 파이썬에서는 매개변수에 실제 인수가 **대입**됨. 복사가 아니라 n과 x가 참조하는 곳이 같음. 그러나 함수 종료 시에는 둘이 달랐음. 이는 바로 int가 이뮤터블 타입이기 때문임. n이 시작할 때에는 x와 같은 식별번호의 객체를 참조하지만, 값이 업데이트됨에 따라 다른 객체를 참조하게 되어 달라지는 것임. 만약 리스트였다면 매개변수가 변경되면 인수도 바뀌게 됨.

파이썬이 아닌 다른 프로그래밍 언어에서는 실제 인수의 값을 매개변수에 복사하는 **값에 의한 호출(call by value)** 또는 실제 인수의 참조를 매개변수에 복사하여 매개변수가 실제 인수와 같아지는 **참조에 의한 호출(call by reference)** 를 사용함. 파이썬은 둘 다 아닌 중간적인 방식으로 값을 전달하는데, 공식 문서에서는 이를 **객체 참조에 의한 전달(call by object reference)** 라는 용어로 설명함.

정리해보면,

함수의 실행 시작 시점에는 매개변수와 실제 인수는 같은 객체 참조. 함수 내에서 매개변수의 값을 변경하면 인수의 형(type)에 따라 다음과 같이 구됨.

1. 인수가 이뮤터블일 때: 매게변수의 값이 변경되면 다른 객체를 생성하여 그 객체에 대한 참조로 업데이트되고, 이에 따라 매개변수의 값을 변경해도 호출하는 쪽의 실제 인수에는 영향 없음
2. 인수가 뮤터블일 때: 함수 내에서 매개변수의 값을 변경하면 객체 자체 업데이트. 즉 실제 인수의 값도 바뀜

[Do it! 실습 2C-6] 리스트에서 임의의 원솟값을 업데이트하기

In [45]:
def change(lst, idx, val):
    lst[idx] = val
    
x = [11, 22, 33, 44, 55]
print('x =', x)

index = int(input('업데이트할 인덱스를 선택하세요.: '))
val = int(input('새로운 값을 입력하세요.: '))

change(x, index, val)
print(f'x = {x}')

x = [11, 22, 33, 44, 55]
업데이트할 인덱스를 선택하세요.: 2
새로운 값을 입력하세요.: 99
x = [11, 22, 99, 44, 55]


위에서 리턴하지 않았어도 함수 내에서 리스트의 값을 변경하도록 구성됨.

## 소수 나열하기

소수(prime number)는 자신과 1 이외의 정수로 나누어 떨어지지 않는 정수. 나누어 떨어지는 정수가 하나 이상 존재하면 그 수는 합성수(composite number)임. 1,000 이하의 모든 소수를 나열하는 프로그램을 짜보자

[Do it! 실습 2-8] 1,000 이하의 소수를 나열하기

In [2]:
counter = 0

for i in range(2, 1001):
    for j in range(2, i):
        counter += 1
        if i % j == 0:
            break
    else:
        print(i)
        
print(f'나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
나눗셈을 실행한 횟수: 78022


n이 2와 3으로 나누어 떨어지지 않으면, 2와 3으로 이루어진 6으로도 나누어 떨어지지 않음. 즉 모든 숫자에 대해 나누어 떨어지는지 확인하기보다는 n보다 작은 소수로 나누어 떨어지는지만 확인하여도 소수인지 알 수 있음. 이 원리를 적용하여 알고리즘을 개선해보자

[Do it! 실습 2-9] 1,000이하의 소수를 나열하기(알고리즘 개선 1)

In [5]:
counter = 0
ptr = 0
prime = [None] * 500               # 미리 초기화

prime[ptr] = 2                     # 2는 미리 넣어두기
ptr += 1                           # 2에 대한 포인터

for i in range(3, 1001, 2):        # 짝수는 어차피 소수가 아님(2제외) 홀수에 대해서만 계산
    for j in range(1, ptr):
        counter += 1
        if i % prime[j] == 0:      # 이미 홀수라서 2 이후부터 나누면 됨.
            break
    else:
        prime[ptr] = i             # 소수를 prime 배열에 넣고
        ptr += 1                   # 포인터 하나 뒤로
        
for i in range(ptr):
    print(prime[i])
        
print(f'나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
나눗셈을 실행한 횟수: 14622


거의 1/7로 줄었음

개선 방법이 하나 더 있음. 기존에 나도 알고 있던 방법인데, 어떤 수의 약수는 제곱근을 기준으로 대칭성을 가짐. 즉 소수인지 검사할 시에는 제곱근 이하의 값에 대해서만 수행하면 됨.

[Do it! 실습 2-10] 1,000이하의 소수를 나열하기(알고리즘 개선 2)

In [14]:
counter = 0
ptr = 0
prime = [None] * 500               # 미리 초기화

prime[ptr] = 2                     # 2는 미리 넣어두기
ptr += 1                           # 2에 대한 포인터

prime[ptr] = 3                     # 3는 미리 넣어두기
ptr += 1                           # 3에 대한 포인터

for i in range(5, 1001, 2):        # 짝수는 어차피 소수가 아님(2제외) 홀수에 대해서만 계산
    j = 1
    while prime[j] * prime[j] <= i:
        counter += 2               # 연산이 2개임
        if i % prime[j] == 0:      # 이미 홀수라서 2 이후부터 나누면 됨.
            break
        j += 1
    else:
        counter += 1
        prime[ptr] = i             # 소수를 prime 배열에 넣고
        ptr += 1                   # 포인터 하나 뒤로
        
for i in range(ptr):
    print(prime[i])
        
print(f'나눗셈을 실행한 횟수: {counter}')

2
3
5
7
11
13
17
19
23
29
31
37
41
43
47
53
59
61
67
71
73
79
83
89
97
101
103
107
109
113
127
131
137
139
149
151
157
163
167
173
179
181
191
193
197
199
211
223
227
229
233
239
241
251
257
263
269
271
277
281
283
293
307
311
313
317
331
337
347
349
353
359
367
373
379
383
389
397
401
409
419
421
431
433
439
443
449
457
461
463
467
479
487
491
499
503
509
521
523
541
547
557
563
569
571
577
587
593
599
601
607
613
617
619
631
641
643
647
653
659
661
673
677
683
691
701
709
719
727
733
739
743
751
757
761
769
773
787
797
809
811
821
823
827
829
839
853
857
859
863
877
881
883
887
907
911
919
929
937
941
947
953
967
971
977
983
991
997
나눗셈을 실행한 횟수: 3774


위처럼 78022 -> 14622 -> 3774번의 연산으로 줄어든 것을 알 수 있음. 다만 가독성 자체는 많이 떨어지는 것을 확인할 수 있음. 효율과 가독성을 적절히 고려하여 코드를 짜야할 듯.

### list.copy(), copy.deepcopy()

copy는 얕은 복사, deepcopy는 깊은 복사임.

In [16]:
a = [[1,2,3], ['x', 'y']]
b = a
print(a, b)

b[0][1] = 10
print(a, b)

[[1, 2, 3], ['x', 'y']] [[1, 2, 3], ['x', 'y']]
[[1, 10, 3], ['x', 'y']] [[1, 10, 3], ['x', 'y']]


In [17]:
a = [[1,2,3], ['x', 'y']]
b = a.copy()
print(a, b)

b[0][1] = 10
print(a, b)

[[1, 2, 3], ['x', 'y']] [[1, 2, 3], ['x', 'y']]
[[1, 10, 3], ['x', 'y']] [[1, 10, 3], ['x', 'y']]


In [18]:
import copy

a = [[1,2,3], ['x', 'y']]
b = copy.deepcopy(a)
print(a, b)

b[0][1] = 10
print(a, b)

[[1, 2, 3], ['x', 'y']] [[1, 2, 3], ['x', 'y']]
[[1, 2, 3], ['x', 'y']] [[1, 10, 3], ['x', 'y']]


깊은 복사를 수행할 때에만, 기존 변수의 참조 객체 내용이 변경되지 않음.

# 마치며..

이번 챕터에서 학습한 내용이야 파이썬의 기초적인 부분이라 새로운 내용이 많지는 않았지만, 이 부분을 명확하게 다시 짚고 넘어가야 이 다음을 잘 해결할 수 있을 것으로 보임. 검색 알고리즘부터 잘 이해하고 공부해보자!!